File: Java/OPSJLib/src/ops/archiver/OPSArchiverOut.java
/**
*
* Copyright (C) 2006-2009 Anton Gravestam.
*
* This file is part of OPS (Open Publish Subscribe).
*
* OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.

* OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
*/

package ops.archiver;

import configlib.ArchiverInOut;
import configlib.Serializable;
import java.io.IOException;
import java.util.List;
import ops.OPSObject;
import ops.WriteByteBuffer;

/**
 *
 * @author angr
 */
public class OPSArchiverOut implements ArchiverInOut
{
    WriteByteBuffer writeBuf;
//
//    public OPSArchiverOut()
//    {
//        writeBuf = new WriteByteBuffer();
//    }

    public OPSArchiverOut(WriteByteBuffer buf)
    {
        this.writeBuf = buf;
    }
    public byte[] getBytes()
    {
        return writeBuf.getBytes();
    }

    public int inout(String name, int v) throws IOException
    {
        writeBuf.write(v);
        return v;
    }

    public long inout(String name, long v) throws IOException
    {
        writeBuf.write(v);
        return v;
    }

    public byte inout(String name, byte v) throws IOException
    {
        writeBuf.write(v);
        return v;
    }

    public short inout(String name, short v) throws IOException
    {
        writeBuf.write(v);
        return v;
    }

    public float inout(String name, float v) throws IOException
    {
        writeBuf.write(v);
        return v;
    }

    public boolean inout(String name, boolean v) throws IOException
    {
        writeBuf.write(v);
        return v;
    }

    public String inout(String name, String v) throws IOException
    {
        writeBuf.write(v);
        return v;
    }

    public double inout(String name, double v) throws IOException
    {
        writeBuf.write(v);
        return v;
    }

    public Serializable inout(String name, Serializable v) throws IOException
    {
        writeBuf.write(((OPSObject)v).getTypesString());
        v.serialize(this);
        return v;
    }

   
    public List<Integer> inoutIntegerList(String name, List<Integer> v) throws IOException
    {
        writeBuf.writeintArr(v);
        return v;
    }

    public List<Long> inoutLongList(String name, List<Long> v) throws IOException
    {
        writeBuf.writelongArr(v);
        return v;
    }

    public List<Byte> inoutByteList(String name, List<Byte> v) throws IOException
    {
        writeBuf.writebyteArr(v);
        return v;
    }

    public List<Short> inoutShortList(String name, List<Short> v) throws IOException
    {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    public List<Float> inoutFloatList(String name, List<Float> v) throws IOException
    {
        writeBuf.writefloatArr(v);
        return v;
    }

    public List<Boolean> inoutBooleanList(String name, List<Boolean> v) throws IOException
    {
        writeBuf.writebooleanArr(v);
        return v;
    }

    public List<String> inoutStringList(String name, List<String> v) throws IOException
    {
        writeBuf.writestringArr(v);
        return v;
    }

    public List<Double> inoutDoubleList(String name, List<Double> v) throws IOException
    {
        writeBuf.writedoubleArr(v);
        return v;
    }

    public List inoutSerializableList(String name, List v) throws IOException
    {
        writeBuf.write(v.size());
        for (int i = 0; i < v.size(); i++)
        {
            writeBuf.write(((OPSObject)v.get(i)).getTypesString());
            ((Serializable)v.get(i)).serialize(this);
        }
        return v;
    }

}

File: C#/Ops/Archiver/OPSArchiverOut.cs
///////////////////////////////////////////////////////////
//  OPSArchiverOut.cs
//  Implementation of the Class OPSArchiverOut
//  Created on:      12-nov-2011 09:25:32
//  Author:
///////////////////////////////////////////////////////////

using System.Collections;
using System.Collections.Generic;

namespace Ops
{
	public class OPSArchiverOut : IArchiverInOut 
    {
		internal WriteByteBuffer writeBuf;

		public OPSArchiverOut(WriteByteBuffer buf)
        {
            this.writeBuf = buf;
        }

        public bool IsOut()
        {
            return true;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public int Inout(string name, int v)
        {
            writeBuf.Write(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public long Inout(string name, long v)
        {
            writeBuf.Write(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public byte Inout(string name, byte v)
        {
            writeBuf.Write(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public short Inout(string name, short v)
        {
            writeBuf.Write(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public float Inout(string name, float v)
        {
            writeBuf.Write(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public bool Inout(string name, bool v)
        {
            writeBuf.Write(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public string Inout(string name, string v)
        {
            writeBuf.Write(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public double Inout(string name, double v)
        {
            writeBuf.Write(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public ISerializable Inout(string name, ISerializable v)
        {
            writeBuf.Write(((OPSObject)v).GetTypesString());
            v.Serialize(this);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public List<bool> InoutBooleanList(string name, List<bool> v)
        {
            writeBuf.WriteBooleanArr(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public List<byte> InoutByteList(string name, List<byte> v)
        {
            writeBuf.WriteByteArr(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public List<double> InoutDoubleList(string name, List<double> v)
        {
            writeBuf.WriteDoubleArr(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public List<float> InoutFloatList(string name, List<float> v)
        {
            writeBuf.WriteFloatArr(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public List<int> InoutIntegerList(string name, List<int> v)
        {
            writeBuf.WriteIntArr(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public List<long> InoutLongList(string name, List<long> v)
        {
            writeBuf.WriteLongArr(v);
            return v;
        }

        // NB! we assume that the object is a List<X> where X implements ISerializable.
        public IList InoutSerializableList(string name, IList v)
        {
            writeBuf.Write(((IList)v).Count);
            foreach (ISerializable obj in (IList)v)
            {
                writeBuf.Write(((OPSObject)obj).GetTypesString());
                obj.Serialize(this);
            }
            return v;
        }

        // NB! we assume that the object is a List<T> where T implements ISerializable.
        public IList InoutSerializableList<T>(string name, IList v)
        {
            return InoutSerializableList(name, v);
        }

        /// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public List<short> InoutShortList(string name, List<short> v)
        {
            writeBuf.WriteShortArr(v);
            return v;
        }

		/// 
		/// <param name="name"></param>
		/// <param name="v"></param>
		public List<string> InoutStringList(string name, List<string> v)
        {
            writeBuf.WriteStringArr(v);
            return v;
        }

	}

}